pallets/pallet-governance/src/lib.rs

```diff
@@ -0,0 +1,407 @@
+		AlreadyVoted,
+		/// Error wouldn't like to happen.
+		ImpossibleReach,
+	}
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(crate) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// Registered the person who made the proposal.
+		ProposalRegistered { who: T::AccountId, call: <T as Config>::RuntimeCall },
+
+		/// Registered the vote result.
+		VoteCasted { who: T::AccountId, proposal: ProposalId, approve: bool },
+
+		/// Rotated the council members.
+		AuthorityRotated { new_council: BTreeSet<T::AccountId> },
+
+		/// Registered the passed proposal.
+		ProposalPassed {
+			id: ProposalId,
+			call: <T as Config>::RuntimeCall,
+			result: DispatchResultWithPostInfo,
+		},
+
+		/// Registered the rejected proposal.
+		ProposalRejected { id: ProposalId, reason: RejectReason },
+	}
+
+	/// Stores the next Proposal ID should be.
+	#[pallet::storage]
+	pub type NextProposalId<T: Config> = StorageValue<_, ProposalId, ValueQuery>;
+
+	/// Stores the proposal id related the call.
+	#[pallet::storage]
+	#[pallet::getter(fn proposals)]
+	pub type Proposals<T: Config> = StorageMap<_, Blake2_128Concat, ProposalId, EncodedCall>;
+
+	/// Stores the proposal IDs that will expire at a black.
+	#[pallet::storage]
+	#[pallet::getter(fn expiry)]
+	pub type Expiry<T: Config> =
+		StorageMap<_, Blake2_128Concat, BlockNumberFor<T>, BTreeSet<ProposalId>>;
+
+	/// Stores the vote results of the proposal id.
+	#[pallet::storage]
+	#[pallet::getter(fn votes)]
+	pub type Votes<T: Config> =
+		StorageMap<_, Blake2_128Concat, ProposalId, CastedVotes<T::AccountId>>;
+
+	/// Stores the current council members.
+	#[pallet::storage]
+	#[pallet::getter(fn authorities)]
+	pub type CurrentAuthorities<T: Config> = StorageValue<_, BTreeSet<T::AccountId>, ValueQuery>;
+
+	/// Stores the resolved proposals.
+	#[pallet::storage]
+	#[pallet::getter(fn resolve)]
+	pub type ProposalsToResolve<T: Config> =
+		StorageValue<_, BTreeSet<(ProposalId, bool)>, ValueQuery>;
+
+	#[pallet::genesis_config]
+	#[derive(frame_support::DefaultNoBound)]
+	pub struct GenesisConfig<T: Config> {
+		initial_authorities: Vec<T::AccountId>,
+	}
+
+	#[pallet::genesis_build]
+	impl<T: Config> BuildGenesisConfig for GenesisConfig<T> {
+		fn build(&self) {
+			// Set CurrentAuthorities storage with initial_authorities.
+			Pallet::<T>::rotate_authorities(self.initial_authorities.clone());
+		}
+	}
+
+	#[pallet::origin]
+	pub type Origin = RawOrigin;
+
+	/// The raw origin enum for this pallet.
+	#[derive(PartialEq, Eq, Clone, RuntimeDebug, Encode, Decode, TypeInfo, MaxEncodedLen)]
+	pub enum RawOrigin {
+		GovernanceApproval,
+	}
+
+	#[pallet::pallet]
+	#[pallet::without_storage_info]
+	pub struct Pallet<T>(_);
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+		fn on_finalize(block_number: BlockNumberFor<T>) {
+			// Check the expiry
+			if let Some(expired_proposals) = Expiry::<T>::get(block_number) {
+				let _ = expired_proposals.iter().map(|proposal_id| {
```

Comment:
iterator maps are lazily evaluated. i.e. the code will execute only when consuming the iterator, i.e. your code here won't do anything.
Also map is for transforming an input into something else. 
use for_each if you want to do something with no return value.

Reply:
This cleanup code is reused after proposals are resolved (Passed or rejected), and expires, so this code can be refactored out into its own function to avoid duplication

Reply:
Also should deposit a event here